packages/@lwc/engine/src/framework/base-lightning-element.ts

```diff
@@ -304,47 +302,11 @@ function BaseLightningElementConstructor(this: LightningElement) {
 // HTML Element - The Good Parts
 BaseLightningElementConstructor.prototype = {
     constructor: BaseLightningElementConstructor,
-    dispatchEvent(event: Event): boolean {
+    dispatchEvent(): boolean {
         const elm = getLinkedElement(this);
-        const vm = getComponentVM(this);
-
-        if (process.env.NODE_ENV !== 'production') {
-            if (arguments.length === 0) {
-                throw new Error(
-                    `Failed to execute 'dispatchEvent' on ${getComponentAsString(
-                        this
-                    )}: 1 argument required, but only 0 present.`
-                );
-            }
-            if (!(event instanceof GlobalEvent)) {
-                throw new Error(
-                    `Failed to execute 'dispatchEvent' on ${getComponentAsString(
-                        this
-                    )}: parameter 1 is not of type 'Event'.`
-                );
-            }
-
-            const { type: evtName } = event;
-            assert.isFalse(
-                isBeingConstructed(vm),
-                `this.dispatchEvent() should not be called during the construction of the custom element for ${getComponentAsString(
-                    this
-                )} because no one is listening for the event "${evtName}" just yet.`
-            );
-
-            if (!/^[a-z][a-z0-9_]*$/.test(evtName)) {
-                logError(
-                    `Invalid event type "${evtName}" dispatched in element ${getComponentAsString(
-                        this
-                    )}. Event name must ${[
-                        '1) Start with a lowercase letter',
-                        '2) Contain only lowercase letters, numbers, and underscores',
-                    ].join(' ')}`,
-                    elm
-                );
-            }
-        }
-        return dispatchEvent.call(elm, event);
+        // Typescript does not like it when you treat the `arguments` object as an array
+        // @ts-ignore type-mismatch;
+        return dispatchEvent.apply(elm, arguments);
```

Comment:
arguments should be sliced, as we do everywhere.

Reply:
> arguments should be sliced, as we do everywhere.

This is not the case, none of the methods on `BaseLightningElementConstructor` nor in `restrictions.ts` slice the arguments today ([`setAttributeNS`](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine/src/framework/base-lightning-element.ts#L399), [`removeAttribute`](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine/src/framework/base-lightning-element.ts#L415), ...). 

From a performance perspective, until recently leaking the `arguments` object to another function caused the javascript engine to deoptimize the invocation, this is not the case today at least on V8 ([source](https://medium.com/the-node-js-collection/get-ready-a-new-v8-is-coming-node-js-performance-is-changing-46a63d6da4de#3284)). When running the benchmark locally, `Array.prototype.slice` is by far the slowest way to pass the arguments object around. Finally, it's important to note that on all the javascript engines passing each argument individually is always faster than manipulating or passing the `arguments` object. 

For the reasons highlighted above and also in order to get rid of all the `//@ts-ignore`, I am thinking about removing the `arguments` references in the `BaseLightningElementConstructor` in favor of direct argument references. 

```ts
// Original
dispatchEvent(_event: Event): boolean {
    const elm = getLinkedElement(this);
    // @ts-ignore type-mismatch
    return dispatchEvent.apply(elm, arguments);
}

// Actual
dispatchEvent(event: Event): boolean {
    const elm = getLinkedElement(this);
    return dispatchEvent.call(elm, event);
}
```

@caridy How does that sound to you?

Reply:
Discussed with @caridy, we want to keep leaking the `arguments` for now, because DOM APIs rely on the passed number of arguments to report specific error messages.

```
function dispatchWithEvent(event: Event) {
    const elm = getLinkedElement(this);
    return dispatchEvent.call(elm, event);
}

dispatchWithEvent(); // Uncaught TypeError: Failed to execute 'dispatchEvent' on 'EventTarget': parameter 1 is not of type 'Event'.

function dispatchWithArguments() {
    const elm = getLinkedElement(this);
    return dispatchEvent.apply(elm, arguments);
}

dispatchWithArguments(); // Uncaught TypeError: Failed to execute 'dispatchEvent' on 'EventTarget': 1 argument required, but only 0 present.
```